File: week1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // Prompt for card number
    long number = get_long("Number: ");

    // Calculate Checksum
    long copy = number;
    long sum = 0;
    long x = 0;
    long remain = copy % 10;

    // 1st step - sum of the digits of (every other digit of card * 2)
    while (copy > 1)
    {
        copy = copy - remain;
        copy = copy / 10;
        remain = copy % 10;
        remain = remain * 2;

        if (remain >= 10)
        {
            sum = sum + 1 + remain % 10;
        }
        else
        {
            sum = sum + remain;
        }
        remain = remain / 2;
        copy = copy - remain;
        copy = copy / 10;
        remain = copy % 10;
    }

    // 2nd step - add the rest of the digits
    copy = number;
    while (copy > 1)
    {
        remain = copy % 10;
        sum = sum + remain;
        copy = copy - remain;
        copy = copy / 10;
        remain = copy % 10;
        copy = copy - remain;
        copy = copy / 10;
    }

    // Check whether it passes Checksum
    int count = 0;
    // Output - if it passes Checksum
    if (sum % 10 == 0)
    {
        while (number > 100)
        {
            number = number / 10;
            count = count + 1;
        }

        //Checks type of card
        if ((number == 34 || number == 37) && count == 13)
        {
            printf("AMEX\n");
        }
        else if ((number == 51 || number == 52 || number == 53 || number == 54 || number == 55) && count == 14)
        {
            printf("MASTERCARD\n");
        }
        else if (number / 10 == 4 && (count == 11 || count == 14))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // Output - if it doesn't pass Checksum
    else
    {
        printf("INVALID\n");
    }
    
}

File: week6/pset6/credit/credit.py
from cs50 import get_int

# Prompt for number
number = get_int("Number: ")

# Checksum
copy = number
digit_sum = 0
x = 0
remain = copy % 10

# 1 - Sum of the digits of (every other digit of card * 2)
while copy > 1:
    copy = copy - remain
    copy = copy / 10
    remain = copy % 10
    remain = remain * 2

    if remain >= 10:
        digit_sum = digit_sum + 1 + remain % 10
    else:
        digit_sum = digit_sum + remain

    remain = remain / 2
    copy = copy - remain
    copy = copy / 10
    remain = copy % 10

# 2 - Add the rest of the digits
copy = number
while copy > 1:
    remain = copy % 10
    digit_sum = digit_sum + remain
    copy = copy - remain
    copy = copy / 10
    remain = copy % 10
    copy = copy - remain
    copy = copy / 10

# Check whether it passes Checksum
count = 0

# CASE I - Pass
if digit_sum % 10 == 0:
    while number > 100:
        number = number // 10
        count += 1

    # Check type of card
    if number == 34 or number == 37:
        if count == 13:
            print("AMEX")

    elif number == 51 or number == 52 or number == 53 or number == 54 or number == 55:
        if count == 14:
            print("MASTERCARD")

    elif number // 10 == 4:
        if count == 11 or count == 14:
            print("VISA")

    else:
        print("INVALID")

# CASE II - Not pass
else:
    print("INVALID")